Concurrent search provider

fpichler opened the pull request:
Solving Memory Consistancy Errors. See CELDEV-242

Fortunately mostly no missing synchronizations, yet using not thread safe HashMap and HashSet let to MCE.

fpichler committed:
CELDEV-242

fpichler committed:
change checkArgument to checkState

fpichler committed:
update celements-core dependency to 1.138

msladek commented on src/main/java/com/xpn/xwiki/plugin/lucene/searcherProvider/SearcherProvider.java:
```diff
@@ -38,82 +41,72 @@
   private Set<Thread> connectedThreads;
 
-  private HashMap<Thread, Set<SearchResults>> connectedSearchResultsMap;
+  private ConcurrentMap<Thread, Set<SearchResults>> connectedSearchResultsMap;
```
You are using Thread as key, but what does Thread.hashCode() do? Consider using Thread.getId() as key instead.

msladek commented on src/main/java/com/xpn/xwiki/plugin/lucene/searcherProvider/SearcherProvider.java:
```diff
@@ -38,82 +41,72 @@
+      checkState(!markToClose, "you may not connect to a SearchProvider marked to close.");
+      LOGGER.debug("connect searcherProvider [{}] to [{}].", System.identityHashCode(this),
+          Thread.currentThread());
       connectedThreads.add(Thread.currentThread());
```
`checkState(!markToClose)` and `connectedThreads.add(Thread.currentThread())` should be in a synchronized block, otherwise a thread could still get connected after the provider was marked to close

msladek commented on src/main/java/com/xpn/xwiki/plugin/lucene/searcherProvider/SearcherProvider.java:
```diff
@@ -38,82 +41,72 @@
-      this.markToClose = true;
+    if (!isMarkedToClose()) {
+      LOGGER.debug("markToClose searcherProvider [{}].", System.identityHashCode(this));
+      markToClose = true;
```
`markToClose = true` should be in a synchronized block, see comment above for `connect()`

msladek commented on src/main/java/com/xpn/xwiki/plugin/lucene/searcherProvider/SearcherProvider.java:
```diff
@@ -38,82 +41,72 @@
-  private boolean markToClose;
+  private volatile boolean isClosed;
 
   private Set<Thread> connectedThreads;
```
Why is this seperate `Set<> connectedThreads` required? Adding an entry directly to `ConcurrentMap<> connectedSearchResultsMap` upon connect() would be the same logically, easier to read and maintain.

msladek commented on src/main/java/com/xpn/xwiki/plugin/lucene/searcherProvider/SearcherProvider.java:
```diff
@@ -156,17 +149,14 @@ protected void finalize() throws Throwable {
-      connectedSearchResultsMap.put(Thread.currentThread(), new HashSet<SearchResults>());
-    }
+    connectedSearchResultsMap.putIfAbsent(Thread.currentThread(), Collections.newSetFromMap(
+        new ConcurrentHashMap<SearchResults, Boolean>()));
```
I think the inserted set can be a regular HashSet since it is only ever accessed by one thread.

fpichler committed:
change access to currentThreadSet in cleanUpSearchResults to use getter
getConnectedSearchResultsForCurrentThread

fpichler commented:
We possibly could use HashSet. Yet, I think we would need to synchronize all read and change access to the HashSet e.g. connectSearchResults , cleanUpSearchResults. Why? Because do not have any guarantee, that this one thread accessing it, will be always executed on the same cpu core.
-> Thus I prefer using Concurrent-Set instead of synchronizing the accesses to the Set.

msladek commented:
If the thread execution switches to another core, then surely the relevant parts of the core cache will be switched with it. in this case, the view of the set will be correct even without any synchronization. am I missing something?

fpichler committed:
replacing Thread dependent concurrent hash set to HashSet

fpichler commented:
ok, convinced. :-) Thus I changed...

msladek commented:
Okay ;) what about my other 3 remarks?

fpichler committed:
replace Thread as key with Thread.getId()

fpichler committed:
adding missing synchronizations

fpichler commented:
sorry, I missed out the other two. :-( now I reworked according to your comments. And added a javadoc explaining the guarantee of the SearchProvider in general and connect-method in special:
connect() is implemented with a fail-fast behavior. It may happen that a thread
 succeeds to connect on a markedToClose SearcherProvider. The guarantee is, that the
 SearcherProvider will not close the connected lucene searchers before not all threads
 finished with theirs SearcherResults and disconnected.

msladek commented:
I see, yet I find this behaviour definition somewhat unintuitive and potentially confusing: wrong usage could throw IllegalStateExceptions or remain unnoticed if you're 'lucky'. I would prefer consistent fast failure.

The synchronized(this) block you've added doesn't affect the program because the AtomicBoolean isn't synchronized on this.

fpichler commented:
The synchronized(this) block is there to ensure that the searchers do not get closed if "connect" successfully passes the isMarkedToClose check before the thread is added to connectedThreads. It is against closeIfIdle

The Atomic Boolean is just to ensure that not multiple threads get into the if block of markToClose. It actually is nice and not forced necessary.

fpichler committed:
add synchronized to martToClose

fpichler commented:
@ebeutler agreed on your fast-failure idea. I changed to implementation. @msladek please review again.

fpichler committed:
fix SearchProviderTest

fpichler committed:
Merge branch 'stable-2.7.2' into concurrentSearchProvider

msladek commented:
looks good now, thanks!
